src/python/fuzzing/coverage_uploader.py

```diff
@@ -81,14 +78,33 @@ def upload_testcases_if_needed(fuzzer_name, testcase_list, testcase_directory):
     if total_testcases >= TESTCASES_PER_DAY:
       return
 
+  # Cap the number of files.
+  testcases_limit = min(len(files_list), TESTCASES_PER_DAY - total_testcases)
+  files_list = files_list[:testcases_limit]
+
   # Upload each batch of tests to its own unique sub-bucket.
   identifier = environment.get_value('BOT_NAME') + str(utils.utcnow())
   gcs_base_url += utils.string_hash(identifier)
 
   list_gcs_url = gcs_base_url + '/' + LIST_FILE_BASENAME
-  if not storage.write_data('\n'.join(testcase_list), list_gcs_url):
+  if not storage.write_data('\n'.join(files_list), list_gcs_url):
     return
 
-  runner.rsync(testcase_directory, gcs_base_url)
-  logs.log('Synced {count} test cases to {gcs_url}'.format(
-      count=len(testcase_list), gcs_url=gcs_base_url))
+  if has_testcases_in_testcase_directory:
+    # Sync everything in |testcase_directory| since it is fuzzer-generated.
+    runner.rsync(testcase_directory, gcs_base_url)
+
+  if has_testcases_in_data_directory:
+    # Sync all fuzzer generated testcase in data bundle directory.
+    runner.rsync(
+        data_directory,
+        gcs_base_url,
+        exclusion_pattern=('"(?!.*{fuzz_prefix})"'.format(
+            fuzz_prefix=testcase_manager.FUZZ_PREFIX)))
+
+    # Sync all possible resource dependencies as a best effort.
+    runner.rsync(
+        data_directory, gcs_base_url, exclusion_pattern='"(?!.*resource)"')
```

Comment:
Which fuzzers is this pattern expected to work for? I'm having trouble thinking of specific fuzzers that write directly to the data directory other than the layout test ones, and I'm not sure that adding this clause helps there.

Reply:
Added explanation.